src/main/java/org/openstreetmap/atlas/checks/validation/intersections/HighwayIntersectionCheck.java

```diff
@@ -0,0 +1,197 @@
+import org.openstreetmap.atlas.geography.atlas.items.Edge;
+import org.openstreetmap.atlas.geography.atlas.items.Node;
+import org.openstreetmap.atlas.geography.atlas.walker.OsmWayWalker;
+import org.openstreetmap.atlas.tags.FordTag;
+import org.openstreetmap.atlas.tags.HighwayTag;
+import org.openstreetmap.atlas.tags.LeisureTag;
+import org.openstreetmap.atlas.tags.Taggable;
+import org.openstreetmap.atlas.tags.WaterwayTag;
+import org.openstreetmap.atlas.tags.annotations.validation.Validators;
+import org.openstreetmap.atlas.utilities.configuration.Configuration;
+
+/**
+ * Flags waterway and power line edge items that are crossed by navigable edges (having way specific
+ * highway tag). If the way is a waterway and the crossing way has {@code ford=yes} or
+ * {@code leisure=slipway} tags, then the crossing is accepted. {@code dam} and {@code weir}
+ * waterways are not checked, those type of ways can cross other highways.
+ *
+ * @author pako.todea
+ * @author brianjor
+ */
+public class HighwayIntersectionCheck extends BaseCheck<Long>
+{
+    private static final long serialVersionUID = -2100623356724302728L;
+    private static final String INSTRUCTION_FORMAT = "The way with id {0,number,#} has invalid intersections with {1}. A navigable way should not share nodes with non-navigable features. Either the way is inproperly tagged or is a combination of what should be two separate ways (highway and the other non-navigable feature).";
+    private static final List<String> FALLBACK_INSTRUCTIONS = Collections
+            .singletonList(INSTRUCTION_FORMAT);
+
+    public HighwayIntersectionCheck(final Configuration configuration)
+    {
+        super(configuration);
+    }
+
+    @Override
+    public boolean validCheckForObject(final AtlasObject object)
+    {
+        return !this.isFlagged(object) && TypePredicates.IS_EDGE.test(object)
+                && ((Edge) object).isMainEdge() && HighwayTag.isCarNavigableHighway(object);
+    }
+
+    @Override
+    protected Optional<CheckFlag> flag(final AtlasObject object)
+    {
+        final Edge edge = (Edge) object;
+        final OsmWayWalker wayWalker = new OsmWayWalker(edge);
+        final Set<Edge> wayEdges = wayWalker.collectEdges();
+        final Stream<Edge> edgesConnectedToWay = wayEdges.stream().map(Edge::connectedEdges)
+                .flatMap(Set::stream)
+                .filter(connectedEdge -> !this.hasSameOSMId(connectedEdge, edge));
+
+        final Set<Edge> invalidconnectingEdges = edgesConnectedToWay
+                .filter(connectedEdge -> TagPredicates.IS_POWER_LINE.test(connectedEdge)
+                        || !FordTag.isYes(edge))
+                .filter(connectedEdge -> TagPredicates.IS_POWER_LINE.test(connectedEdge)
+                        || this.isWaterwayToCheck(connectedEdge))
+                .filter(connectedEdge -> TagPredicates.IS_POWER_LINE.test(connectedEdge)
+                        || !this.isSlipway(edge))
```

Comment:
It seems like there is a lot of redundancy here, caused by having three separate filter calls.